[tests-only] [full-ci] Added acceptance tests for move and rename activities

kiranparajuli589 opened the pull request:
see issue https://github.com/owncloud/activity/issues/1019

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiActivity-master-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/activity/2259/25/1

ownclouders commented:
:boom: Acceptance tests pipeline <strong>webUISharingExt-master-firefox-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/activity/2261/24/1

ownclouders commented:
:boom: Acceptance tests pipeline <strong>webUISharingExt-master-firefox-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/activity/2263/24/1

JammingBen commented:
@kiranparajuli589 Fyi I pushed and squashed the source branch, hence the conflict here. Just ping if you need help resolving :)

kiranparajuli589 commented:
> @kiranparajuli589 Fyi I pushed and squashed the source branch, hence the conflict here. Just ping if you need help resolving :)

thank you, i rebased it...open for your reviews

JammingBen commented on tests/acceptance/features/apiActivity/list.feature:
```diff
@@ -725,4 +806,229 @@ Feature: List activity
+      | object_type      | /^files$/                                                                                                                                                                                                                 |
+      | typeicon         | /^icon-rename/                                                                                                                                                                                                            |
+      | subject_prepared | /^You renamed <file link=\"%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/&scrollto=FOLDER\" id=\"\">FOLDER<\/file> to <file link=\"%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/PARENT\" id=\"\d+\">PARENT<\/file>$/ |
+    And the activity number 1 of user "Brian" should match these properties:
```
For Brian, the name of the share (= the mountpoint) doesn't change in that scenario. Therefore he shouldn't see an activity about it. Although this is still a bug in the source branch I need to fix.

JammingBen replied:
> Therefore he shouldn't see an activity about it.

I updated the source branch, this should work now.

kiranparajuli589 replied:
will this be fixed in this `feat.` PR?

JammingBen replied:
`And the activity number 1 of user "Brian" should match these properties:` should now be something like `And user "Brian" does not see any activities`.

kiranparajuli589 replied:
user Brian gets two activities data at this step. one with type `shared` and another `file_created`
```console
  Scenario: check activity after renaming a shared folder                                   # /home/kiran/www/core/apps-external/activity/tests/acceptance/features/apiActivity/list.feature:812
    Given these users have been created with default attributes and without skeleton files: # FeatureContext::theseUsersHaveBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()
      | username |
      | Alice    |
      | Brian    |
    And user "Alice" has created folder "FOLDER"                                            # FeatureContext::userHasCreatedFolder()
    And user "Alice" has shared folder "/FOLDER" with user "Brian"                          # FeatureContext::userHasSharedFileWithUserUsingTheSharingApi()
    And user "Brian" has accepted share "/FOLDER" offered by user "Alice"                   # FeatureContext::userHasReactedToShareOfferedBy()
    When user "Alice" moves folder "/FOLDER" to "/PARENT" using the WebDAV API              # FeatureContext::userMovesFileUsingTheAPI()
    Then the activity number 1 of user "Alice" should match these properties:               # ActivityContext::activityWithIndexShouldMatch()
      | type             | /^file_renamed$/                                                                                                                                                                                                          |
      | user             | /^Alice$/                                                                                                                                                                                                                 |
      | affecteduser     | /^Alice$/                                                                                                                                                                                                                 |
      | app              | /^files$/                                                                                                                                                                                                                 |
      | subject          | /^renamed_self$/                                                                                                                                                                                                          |
      | object_name      | /^\/PARENT/                                                                                                                                                                                                               |
      | object_type      | /^files$/                                                                                                                                                                                                                 |
      | typeicon         | /^icon-rename/                                                                                                                                                                                                            |
      | subject_prepared | /^You renamed <file link=\"%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/&scrollto=FOLDER\" id=\"\">FOLDER<\/file> to <file link=\"%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/PARENT\" id=\"\d+\">PARENT<\/file>$/ |
    And user "Brian" should not see any activities                                          # ActivityContext::userShouldNotSeeAnyActivities()
      │ /home/kiran/www/core/apps-external/activity/tests/acceptance/features/bootstrap/ActivityContext.php:110:
      │ array(2) {
      │   [0] =>
      │   array(17) {
      │     'activity_id' =>
      │     int(778)
      │     'type' =>
      │     string(6) "shared"
      │     'user' =>
      │     string(5) "Alice"
      │     'affecteduser' =>
      │     string(5) "Brian"
      │     'app' =>
      │     string(13) "files_sharing"
      │     'subject' =>
      │     string(14) "shared_with_by"
      │     'subjectparams' =>
      │     array(2) {
      │       [0] =>
      │       array(2) {
      │         ...
      │       }
      │       [1] =>
      │       array(2) {
      │         ...
      │       }
      │     }
      │     'message' =>
      │     string(0) ""
      │     'messageparams' =>
      │     array(0) {
      │     }
      │     'link' =>
      │     string(55) "http://localhost/core/index.php/apps/files/?dir=/FOLDER"
      │     'object_type' =>
      │     string(5) "files"
      │     'object_id' =>
      │     int(2147484456)
      │     'object_name' =>
      │     string(7) "/FOLDER"
      │     'subject_prepared' =>
      │     string(162) "<user display-name="Alice Hansen">Alice</user> shared <file link="http://localhost/core/index.php/apps/files/?dir=/FOLDER/" id="2147484456">FOLDER</file> with you"
      │     'message_prepared' =>
      │     string(0) ""
      │     'typeicon' =>
      │     string(11) "icon-shared"
      │     'datetime' =>
      │     string(25) "2021-11-03T10:48:06+00:00"
      │   }
      │   [1] =>
      │   array(17) {
      │     'activity_id' =>
      │     int(775)
      │     'type' =>
      │     string(12) "file_created"
      │     'user' =>
      │     string(5) "Brian"
      │     'affecteduser' =>
      │     string(5) "Brian"
      │     'app' =>
      │     string(5) "files"
      │     'subject' =>
      │     string(12) "created_self"
      │     'subjectparams' =>
      │     array(1) {
      │       [0] =>
      │       array(2) {
      │         ...
      │       }
      │     }
      │     'message' =>
      │     string(0) ""
      │     'messageparams' =>
      │     array(0) {
      │     }
      │     'link' =>
      │     string(48) "http://localhost/core/index.php/apps/files/?dir="
      │     'object_type' =>
      │     string(5) "files"
      │     'object_id' =>
      │     int(2147484455)
      │     'object_name' =>
      │     string(0) ""
      │     'subject_prepared' =>
      │     string(99) "You created <file link="http://localhost/core/index.php/apps/files/?dir=/" id="2147484455">/</file>"
      │     'message_prepared' =>
      │     string(0) ""
      │     'typeicon' =>
      │     string(14) "icon-add-color"
      │     'datetime' =>
      │     string(25) "2021-11-03T10:48:05+00:00"
      │   }
      │ }
      │ 
```

JammingBen replied:
Ah okay, makes sense. Either way, we need to make sure here that Brian does not see the rename activity that Alices sees. I don't know the underlying test-implementation, what is `the activity number 1`? Maybe we can check that this activity is not a rename?

phil-davis replied:
I added step definitions that let us check that "User" has no activity entries at all, and that they have no activity entries of a specified type. I adjusted this scenario.

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiActivity-master-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/activity/2270/25/1

JammingBen commented on tests/acceptance/features/apiActivity/list.feature:
```diff
@@ -725,4 +806,229 @@ Feature: List activity
+      | subject_prepared | /^<user display-name=\"Alice Hansen\">Alice<\/user> renamed <file link="%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/FOLDER\&scrollto=textfile0\.txt" id=\"\">FOLDER\/textfile0\.txt<\/file> to <file link="%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/FOLDER\&scrollto=textfile1\.txt\" id=\"\d+\">FOLDER\/textfile1\.txt<\/file>$/ |
+
+
+  Scenario: move a file inside a share
```
It seems this scenario is `move a file into a share`.

The behavior here changed slightly yesterday. For Alice it's correct. However, Brian will see a `created_by` activity instead of `moved_in_share_by`.

kiranparajuli589 replied:
yes noticed that in ci too...i'll push the fix asap

JammingBen commented on tests/acceptance/features/apiActivity/list.feature:
```diff
@@ -725,4 +806,229 @@ Feature: List activity
+      | typeicon         | /^icon-move/                                                                                                                                                                                                        |
+      | subject_prepared | /^<user display-name=\"Alice Hansen\">Alice<\/user> moved <file link=\"%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/FOLDER\&scrollto=textfile0\.txt\" id=\"\d+\">FOLDER\/textfile0\.txt<\/file> into the share$/ |
+
+  Scenario: move a file inside a share (inside Shares folder)
```
Same as above, this seems more like `move a file into a share (inside Shares folder)`.

JammingBen commented on tests/acceptance/features/apiActivity/list.feature:
```diff
@@ -725,4 +806,229 @@ Feature: List activity
+      | subject_prepared | /^<user display-name=\"Alice Hansen\">Alice<\/user> moved <file link=\"%base_url%\/(index\.php\/)?apps\/files\/\?dir=\/Shares\/FOLDER\&scrollto=textfile0\.txt\" id=\"\d+\">Shares\/FOLDER\/textfile0\.txt<\/file> into the share$/ |
+
+
+  Scenario: move a file out of a share
```
The behavior here changed slightly yesterday. For Alice it's correct. However, Brian will see a `deleted_by` activity instead of `moved_in_share_by`.

JammingBen commented on tests/acceptance/features/apiActivity/list.feature:
```diff
@@ -725,4 +806,229 @@ Feature: List activity
+      | Brian    |
+    And user "Alice" has created folder "folder"
+    And user "Alice" has created folder "folder/sub"
+    And user "Alice" has uploaded file with content "file to share" to "/textfile0.txt"
```
This would make it a "move into share" event. `textfile0.txt` must be uploaded to `/folder/textfile0.txt` initially (and then moved to `/folder/sub/textfile0.txt`).

JammingBen commented on tests/acceptance/features/apiActivity/list.feature:
```diff
@@ -725,4 +806,229 @@ Feature: List activity
+    And the administrator has set the default folder for received shares to "Shares"
+    And user "Alice" has created folder "folder"
+    And user "Alice" has created folder "folder/sub"
+    And user "Alice" has uploaded file with content "file to share" to "/textfile0.txt"
```
Same as above -> `textfile0.txt` must be uploaded to `/folder/textfile0.txt` initially.

kiranparajuli589 replied:
fixed

kiranparajuli589 commented:
@JammingBen addressed your reviews. please review again

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiActivity-master-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/activity/2273/25/1

JammingBen commented:
See my comment in https://github.com/owncloud/activity/pull/1022#discussion_r741737092. Besides that, things are looking good IMO!

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiActivity-master-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/activity/2275/25/1

phil-davis commented:
ToDo:

- [ ] add scenarios for moving a resource between two shares (shared to different users)
- [ ] add a few scenarios for when the share is to a group, making sure that all group members see the activity
- [ ] ...?

phil-davis commented:
@JammingBen I cherry-picked the latest commits from #1018 into here, and adjusted the "rename share" scenario so that it now passes.

We could merge this PR into #1018 now. And that would get the acceptance tests that we have so far into the PR.

(Then I have another "moving between shares" scenario locally that does not seem quite right - I can make a new PR for that and we can discuss exactly what the behavior should be).

JammingBen commented:
Why are there changes in `lib/FilesHooks.php`? These are already present in the source branch 🤔

phil-davis commented:
> Why are there changes in `lib/FilesHooks.php`? These are already present in the source branch thinking

That is the cherry-pick that I needed to do. This branch only comes off the 2nd commit of 4 in your branch. They will "disappear" in the merging - it is the same code.

JammingBen commented:
Ahh okay. LGTM then.

> (Then I have another "moving between shares" scenario locally that does not seem quite right - I can make a new PR for that and we can discuss exactly what the behavior should be).

Sounds good!
